037-prototypes/proposal.md

```diff
@@ -0,0 +1,473 @@
+send, and a `type` denoting the prototype to use:
+
+```yaml
+prototypes:
+- name: some-prototype
+  type: registry-image
+  source:
+    repository: vito/some-prototype
+
+jobs:
+- name: run-prototype
+  plan:
+  - run: some-message
+    type: some-prototype
+```
+
+So far this example is not very useful. There aren't many things I can think to
+run that don't take any inputs or configuration or produce any outputs. Let's
+try giving it an input and collecting its resulting output. This can be done
+with `inputs`/`input_mapping` and `outputs`/`output_mapping`:
+
+```yaml
+prototypes:
+- name: oci-image
+  type: registry-image
+  source:
+    repository: vito/oci-image-prototype
+
+resources:
+- name: my-image-src
+  type: git
+  source:
+    uri: https://example.com/my-image-src
+
+- name: my-image
+  type: registry-image
+  source:
+    repository: example/my-image
+    username: some-username
+    password: some-password
+
+jobs:
+- name: build-and-push
+  plan:
+  - get: my-image-src
+  - run: build
+    type: oci-image
+    input_mapping: {context: my-image-src}
+    outputs: [image]
+  - put: my-image
+    params: {image: image/image.tar}
+```
+
+This example also makes use of resources, which are also implemented as
+prototypes with special pipeline pipeline semantics and step syntax. These
+"resource prototypes" are described in [RFC #38][rfc-38].
+
+
+## Out of scope
+
+* [Richer metadata](https://github.com/concourse/concourse/issues/310) - this
+  hasn't gained much traction and probably needs more investigation before it
+  can be incorporated. This should be easy enough to add as a later RFC.
+
+
+## Etymology
+
+* The root of the name comes from [Prototype-based
+  programming](https://en.wikipedia.org/wiki/Prototype-based_programming), a
+  flavor of object-oriented programming, which the concepts introduced in this
+  proposal mirror somewhat closely. There are also some hints of actor-based
+  programming - actors handling messages and emitting more actors - but that
+  felt like more of a stretch.
+
+  This is also an opportunity to frame Concourse as "object-oriented CI/CD," if
+  we want to.
+
+* We've been conflating the terms 'resource' and 'resource type' pretty
+  everywhere which seems like it would be confusing to newcomers. For example,
+  all of our repos are typically named 'X-resource' when really the repo
+  provides a resource *type*.
+
+  Adopting a new name which still has the word 'type' in it feels like it fixes
+  this problem, while preserving 'resource' terminology for its original use
+  case (pipeline resources).
+
+* It's a pun: 'protocol type' => 'prototype'.
+
+* 'Prototype' has a pretty nifty background in aerospace engineering. This is a
+  completely different meaning, but we never use it that way, so it doesn't
+  seem like there should be much cause for confusion.
+
+
+## Open Questions
+
+* Is this terminology unrelatable?
```

Comment:
I find it unrelatable but not ambiguous. Given the importance I perceive in either "dependent-types of prototypes" (i.e. any type implementing some set of messages that relate to each other) or "categories" as described above (which include the high level object and its hypothetically invariant properties).